// @expo-google-fonts/montserrat/index.js
// This file is generated by @expo-google-fonts/generator
// If you want to modify it, go to packages/generator/templates

export * from './useFonts';
export { default as __metadata__ } from './metadata.json';
export const Montserrat_100Thin = require('./100Thin/Montserrat_100Thin.ttf');
export const Montserrat_200ExtraLight = require('./200ExtraLight/Montserrat_200ExtraLight.ttf');
export const Montserrat_300Light = require('./300Light/Montserrat_300Light.ttf');
export const Montserrat_400Regular = require('./400Regular/Montserrat_400Regular.ttf');
export const Montserrat_500Medium = require('./500Medium/Montserrat_500Medium.ttf');
export const Montserrat_600SemiBold = require('./600SemiBold/Montserrat_600SemiBold.ttf');
export const Montserrat_700Bold = require('./700Bold/Montserrat_700Bold.ttf');
export const Montserrat_800ExtraBold = require('./800ExtraBold/Montserrat_800ExtraBold.ttf');
export const Montserrat_900Black = require('./900Black/Montserrat_900Black.ttf');
export const Montserrat_100Thin_Italic = require('./100Thin_Italic/Montserrat_100Thin_Italic.ttf');
export const Montserrat_200ExtraLight_Italic = require('./200ExtraLight_Italic/Montserrat_200ExtraLight_Italic.ttf');
export const Montserrat_300Light_Italic = require('./300Light_Italic/Montserrat_300Light_Italic.ttf');
export const Montserrat_400Regular_Italic = require('./400Regular_Italic/Montserrat_400Regular_Italic.ttf');
export const Montserrat_500Medium_Italic = require('./500Medium_Italic/Montserrat_500Medium_Italic.ttf');
export const Montserrat_600SemiBold_Italic = require('./600SemiBold_Italic/Montserrat_600SemiBold_Italic.ttf');
export const Montserrat_700Bold_Italic = require('./700Bold_Italic/Montserrat_700Bold_Italic.ttf');
export const Montserrat_800ExtraBold_Italic = require('./800ExtraBold_Italic/Montserrat_800ExtraBold_Italic.ttf');
export const Montserrat_900Black_Italic = require('./900Black_Italic/Montserrat_900Black_Italic.ttf');
